tools/reactive-controllers/src/RovingTabindex.ts

```diff
@@ -0,0 +1,304 @@
+        return this._focused;
+    }
+    private _focused = false;
+    get focusInElement(): T {
+        return this.elements[this.focusInIndex];
+    }
+    get focusInIndex(): number {
+        return this.calculateFocusInIndex(this.elements);
+    }
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    isFocusableElement = (_el: T): boolean => true;
+    isEventWithinListenerScope(event: Event): boolean {
+        if (this._listenerScope() === this.host) return true;
+        return event.composedPath().includes(this._listenerScope());
+    }
+    _listenerScope = (): HTMLElement => this.host;
+    private managed = true;
+
+    constructor(
+        host: ReactiveElement,
+        {
+            calculateFocusInIndex,
+            direction,
+            elementEnterAction,
+            elements,
+            isFocusableElement,
+            listenerScope,
+        }: {
+            calculateFocusInIndex?: (_elements: T[]) => number;
+            direction?: DirectionTypes | (() => DirectionTypes);
+            elementEnterAction?: (el: T) => void;
+            elements: () => T[];
+            isFocusableElement?: (el: T) => boolean;
+            listenerScope?: HTMLElement | (() => HTMLElement);
+        } = { elements: () => [] }
+    ) {
+        this.host = host;
+        this.host.addController(this);
+        this._elements = elements;
+        this.isFocusableElement = isFocusableElement || this.isFocusableElement;
+        if (typeof direction === 'string') {
+            this._direction = () => direction;
+        } else if (typeof direction === 'function') {
+            this._direction = direction;
+        }
+        this.calculateFocusInIndex =
+            calculateFocusInIndex || this.calculateFocusInIndex;
+        this.elementEnterAction = elementEnterAction || this.elementEnterAction;
+        if (typeof listenerScope === 'object') {
+            this._listenerScope = () => listenerScope;
+        } else if (typeof listenerScope === 'function') {
+            this._listenerScope = listenerScope;
+        }
+    }
+
+    focus(options?: FocusOptions): void {
+        this.elements[this.currentIndex]?.focus(options);
+    }
+
+    private manageIndexesAnimationFrame = 0;
+
+    clearElementCache(): void {
+        delete this.cachedElements;
+        cancelAnimationFrame(this.manageIndexesAnimationFrame);
+        if (!this.managed) return;
+
+        this.manageIndexesAnimationFrame = requestAnimationFrame(() =>
+            this.manageTabindexes()
+        );
+    }
+
+    setCurrentIndexCircularly(diff: number): void {
+        const { length } = this.elements;
+        let nextIndex = (length + this.currentIndex + diff) % length;
+        while (
+            this.elements[nextIndex] &&
+            !this.isFocusableElement(this.elements[nextIndex])
+        ) {
+            nextIndex = (length + nextIndex + diff) % length;
+        }
+        this.currentIndex = nextIndex;
+    }
```

Comment:
`while`s always make me start thinking defensively about the potential for infinite loops.

What it ever be possible to enter this function with 10 elements that test negative for `isFocusableElement` (either by being disabled, or by the user providing a bad `isFocusableElement` like `return false`)?

To be extra safe, I might limit iterations to `length`, since we never want to loop in more than a circle:

```ts
for (let i = 0; i < length; i++) {
  const index = (length + i + diff) % length;
  if (this.elements[index] && this.isFocusableElement(this.elements[index])) {
    this.currentIndex = index;
    return;
  }
}
```

Less dangerous: could length ever be 0? (x % 0 = NaN)